File: Source/Samples/95_Periodic/code_ascrollbar.cpp
// AScrollBar application source code
#include <Urho3D/AUI/AScrollBar.h>
#include "PeriodicApp.h"

void PeriodicApp::setup_ascrollbar( AWidget *layout )
{
    PODVector<AWidget*> dest;
    layout->SearchWidgetClass( "TBButton", dest );
    for (unsigned ii = 0; ii < dest.Size(); ii++)
        SubscribeToEvent(dest[ii], E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAscrollbarEvent ));

    AWidget *sbx = layout->GetWidget("scrollbardemo");
    if (sbx)  // warning - this will route for all scrollbar instances events into this event handler.
        sbx->SubscribeToEvent( E_WIDGETEVENT, URHO3D_HANDLER(PeriodicApp, HandleAllScrollcontainerEvent ));
}

void PeriodicApp::HandleAllScrollcontainerEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;

    AWidget *sbx = widget->FindWidget("scrollbardemo");  // find our scrollbar
    if ( widget != sbx ) return; // if its not ours, bail

    if (eventData[P_TYPE] == UI_EVENT_TYPE_CHANGED )
    {
        if (widget->GetId() ==  "scrollbardemo" )
        {
            AppLog( "AScrollBar action : " + widget->GetId() + " changed value to " + String (widget->GetValue()));
        }
    }
}

void PeriodicApp::HandleAscrollbarEvent(StringHash eventType, VariantMap& eventData)
{
    using namespace WidgetEvent;
    AWidget* widget = static_cast<AWidget*>(eventData[P_TARGET].GetPtr());
    if ( widget == NULL ) return;

    if (eventData[P_TYPE] == UI_EVENT_TYPE_CLICK)
    {
        if (widget->GetId() == "uiscrollbarcode" )
        {
            AppLog( "AScrollBar support : " + widget->GetId() + " was pressed " );
            ViewCode ( "Sources/code_ascrollbar.cpp", widget->GetParent() );
        }
        if (widget->GetId() ==  "uiscrollbarlayout" )
        {
            AppLog( "AScrollBar support : " + widget->GetId() + " was pressed ");
            ViewCode ( "Scenes/layout_ascrollbar.ui.txt", widget->GetParent() );
        }
    }
}

File: bin/Data/AUI/Periodic/LuaScripts/code_ascrollbar.lua
-- AScrollBar application source code

function setup_uiscrollbar( layout )
    local dest = layout:SearchAWidgetClass( "TBButton" )
    for i, awid in ipairs(dest) do  -- set bulk event handlers on all buttons -- boom!
        SubscribeToEvent(awid, "WidgetEvent", "HandleUiscrollbarEvent")
	end

    local sc1 = layout:GetWidget ("scrollbardemo")
    if sc1 ~= nil then
        SubscribeToEvent(sc1, "WidgetEvent", "HandleUiscrollbarEvent" )
	end
end

function HandleUiscrollbarEvent( eventType, eventData)
    local widget = eventData["Target"]:GetPtr("AWidget")
    if widget == nil then
 		return
	end
    if eventData["Type"]:GetInt() == UI_EVENT_TYPE_CLICK then
        if widget:GetId() == "uiscrollbarcode" then
            AppLog( "AScrollBar support : " .. widget:GetId() .. " was pressed " )
            ViewCode ( "LuaScripts/code_ascrollbar.lua", widget:GetParent())
        end
        if widget:GetId() ==  "uiscrollbarlayout" then
            AppLog( "AScrollBar support : " .. widget:GetId() .. " was pressed ")
            ViewCode ( "Scenes/layout_ascrollbar.ui.txt", widget:GetParent())
        end
    elseif eventData["Type"]:GetInt() == UI_EVENT_TYPE_CHANGED then
        if widget:GetId() == "scrollbardemo" then
            AppLog( "AScrollBar action : " .. widget:GetId() .. " changed value to " .. widget:GetValue())
        end
	else
       	AppLog( "HandleUiscrollbarEvent " .. widget:GetId() .. " got event " .. eventData["Type"]:GetInt())
     end
end

